extensions/ql-vscode/src/data-extensions-editor/extensions-workspace-folder.ts

```diff
@@ -0,0 +1,143 @@
+import { Uri, window, workspace, WorkspaceFolder } from "vscode";
+import { getOnDiskWorkspaceFoldersObjects } from "../common/vscode/workspace-folders";
+import { extLogger } from "../common";
+
+/**
+ * Returns the ancestors of this path in order from furthest to closest (i.e. root of filesystem to parent directory)
+ */
+function getAncestors(uri: Uri): Uri[] {
+  const ancestors: Uri[] = [];
+  let current = uri;
+  while (current.fsPath !== Uri.joinPath(current, "..").fsPath) {
+    ancestors.push(current);
+    current = Uri.joinPath(current, "..");
+  }
+
+  // The ancestors are now in order from closest to furthest, so reverse them
+  ancestors.reverse();
+
+  return ancestors;
+}
+
+function getRootWorkspaceDirectory(): Uri | undefined {
+  // If there is a valid workspace file, just use its directory as the directory for the extensions
+  const workspaceFile = workspace.workspaceFile;
+  if (workspaceFile?.scheme === "file") {
+    return Uri.joinPath(workspaceFile, "..");
+  }
+
+  const workspaceFolders = getOnDiskWorkspaceFoldersObjects();
```

Comment:
Does `getOnDiskWorkspaceFoldersObjects` return only `file:` URIs? If not, you will need to filter out the all the other ones.

Reply:
Yes, that's exactly what `getOnDiskWorkspaceFoldersObjects` does, so this should only return workspace folders which actually exist on-disk.